src/acceleration/BaseQNAcceleration.cpp

```diff
@@ -351,9 +351,14 @@ void BaseQNAcceleration::performAcceleration(
     _preconditioner->update(false, _values, _residuals);
     // apply scaling to V, V' := P * V (only needed to reset the QR-dec of V)
     _preconditioner->apply(_matrixV);
+    if (_preconditioner->updatedWeights() == true ){
+      // If the pre-scaling weights have been updated, then the normal QR2 filter must be run.
+      _qrV.computeQR2 = true;
+    }
 
     if (_preconditioner->requireNewQR()) {
-      if (not(_filter == Acceleration::QR2FILTER)) { //for QR2 filter, there is no need to do this twice
+      if ( ( not(_filter == Acceleration::QR2FILTER) && not(_filter == Acceleration::QR3FILTER) ) || (its < 3 && tWindows == 0) ) { //for QR2 filter, there is no need to do this twice
```

Comment:
Please make "3" a (const) variable, whose name tells us what the variable is doing. And add documentation why the value is 3.